### DISCUSSION TITLE
Welcome to Chalet Discussions!

### DISCUSSION THREAD

**Discussion by rewrking** [chalet-org/chalet#23] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by dimitre** [comment]
Hello, first of all congrats about the project, I've been testing today and it seems incredible.

I would love to know how can I pass -frtti parameter to the compiler.
here is my experimental chalet.json now:
```json
{
	"name": "OF",
	"version": "0.11.3",
	"externalDependencies": {
		"glm" : {
			"repository" : "https://github.com/g-truc/glm"
		},
		"glfw": {
		   "repository": "https://github.com/glfw/glfw"
		   
		}

	 },	

	 "abstracts:*": {
		"language": "C++",
		"settings:Cxx": {
			"cppStandard": "c++17",
			"runtimeTypeInformation": false,
			"defines": [
				" -frtti"
			],
			"includeDirs": [
				"${external:glm}",
				"${external:glfw}",
				"/Volumes/tool/ofw/libs/glew/include",
				"/Volumes/tool/ofw/libs/tess2/include",
				"/Volumes/tool/ofw/libs/utf8/include",
				"/Volumes/tool/ofw/libs/Freeimage/include",
				"/Volumes/tool/ofw/libs/cairo/include/cairo",
				"/Volumes/tool/ofw/libs/curl/include",
				"/Volumes/tool/ofw/libs/fmod/include",
				"/Volumes/tool/ofw/libs/freetype/include",
				"/Volumes/tool/ofw/libs/json/include",
				"/Volumes/tool/ofw/libs/pugixml/include",
				"/Volumes/tool/ofw/libs/rtAudio/include",
				"/Volumes/tool/ofw/libs/uriparser/include"
			],

			"libDirs": [
				"${external:glm}",
				"${external:glfw}",
				"/Volumes/tool/ofw/libs/glew/include"
			],
			"macosFrameworks": [
				"CoreFoundation",
				"CoreAudio",
				"CoreVideo",
				"AppKit",
				"IOKit",
				"Carbon",
				"Metal",
				"AudioToolbox"
			]
		}
	},	 
	"targets": {

		
		"core": {
			"kind": "executable",
			"language": "C++",
			"settings:Cxx": {
				"cppStandard": "c++17",
				"defines": [
					" -frtti"
				],


				"includeDirs": [
					"src/of/libs/openFrameworks",
					"src/of/libs/openFrameworks/3d",
					"src/of/libs/openFrameworks/app",
					"src/of/libs/openFrameworks/communication",
					"src/of/libs/openFrameworks/events",
					"src/of/libs/openFrameworks/gl",
					"src/of/libs/openFrameworks/graphics",
					"src/of/libs/openFrameworks/math",
					"src/of/libs/openFrameworks/sound",
					"src/of/libs/openFrameworks/types",
					"src/of/libs/openFrameworks/utils",
					"src/of/libs/openFrameworks/video"
				]
			},
			"files": "src/of/examples/3d/3DPrimitivesExample/src/**"
		}
	},
	"distribution": {
		"all": {
			"kind": "bundle",
			"buildTargets": "*"
		}
	}
}

```

**Reply by rewrking** [reply]
Thanks for trying it out! It means a lot. There's still lots to iron out, but it's getting there.

To answer your question: set `"runtimeTypeInformation": true` inside `settings:Cxx`, but that just uses rtti implicitly right now -- it's enabled by default in gcc. If you want to explicitly set `-frtti`, use `"compileOptions[toolchain:(name)]": [ "-frtti" ]` where `(name)` can be anything matching your toolchain. Details on that [here](https://www.chalet-work.space/docs/creating-targets?id=property-conditions).

`defines` in this case just refers to CLI macro definitions (`DEBUG`, `UNICODE` or something). They could certainly be validated better.

I may just make `"runtimeTypeInformation": true` explicitly set the flag in the future though...

**Reply by dimitre** [reply]
Outstanding. in fact I am quite happy with Chalet, and I'm trying to build openFrameworks project with it in macOs
https://github.com/openframeworks/openFrameworks
it is a quite complex project (lots of plataforms and variations) but I think your tool is great.

Not sure if it is the best channel but here some initial doubts:
how can I mix c++ and objective-c++ ? or how can I add some compiler parameters like:
```
-DNDEBUG
-fexceptions
-fpascal-strings
-mmacosx-version-min=10.15
-fobjc-arc
-D__MACOSX_CORE__
```

**Reply by rewrking** [reply]
For all the objective-C++ stuff, try to put them into their own staticLibrary target and set the `language` to `Objective-C` (.m files) or `Objective-C++` (.mm files), then link that static library in a C++ executable target. Objective-C isn't really cross-platform, so there's no abstraction properties like with rtti, so you would just add the flags you want to `compilerOptions`:

```json
{
  "abstracts:*": {
    "settings:Cxx": {
      "defines[:macos]": [
        "__MACOSX_CORE"
      ],
      "defines[:!debug]": [ 
        "NDEBUG"
      ],
      "compilerOptions[:macos]": [
        "-mmacos-version-min=10.15"
      ]
    }
  },
  "targets": {
    "my-static-lib": {
      "language": "Objective-C++",
      "kind": "staticLibrary",
      "files": "src/**.mm",
      "settings:Cxx": {
        "condition": "[:macos]",
        "exceptions": true,
        "compilerOptions": [
          "-fpascal-strings",
          "-fobjc-arc"
        ]

      }
    },
    "my-executable": {
      "language": "C++",
      "kind": "executable",
      "files": "src/**.cpp",
      "links[:macos]": [
        "my-static-lib"
      ]
    }
  }
}
```

`-D` options just get translated to `defines`, where you'll want to keep build configurations in mind.

**Reply by dimitre** [reply]
Thank you! here are my results using 0.5.4
```
ERROR: Failed to validate the json file: chalet.json
ERROR: The 'settings:Cxx' object contains an unknown property 'condition': Not allowed.
ERROR: The 'settings:Cxx' object contains an unknown property 'compilerOptions': Not allowed.
ERROR: Instance is too short as per minLength:1
ERROR: The 'app' object contains an unknown property 'links[:macos]': Not allowed.
```

**Reply by rewrking** [reply]
Ah that was my bad. Put condition outside of settings.

**Comment by dimitre** [comment]
One suggestion on the Chalet site : the search results doesn't point to documentation, nothing happens when I click on them.
it would be great to have the entire reference in a PDF file.

now I'm trying to research "macosFrameworks" list. it seems frameworks are not being passed to CXX parameters, but it can be something missing here on my side.

**Reply by rewrking** [reply]
Re: PDF - probably not any time soon, although I'll check out if something is up with the search. It's kind of funky in Safari, but should work fine in chrome/firefox.

For frameworks, this sort of thing should be enough:

```json
"macosFrameworkPaths": [
  "path/to/some/frameworks"
],
"macosFrameworks": [
  "CoreFoundation",
  "CoreAudio",
  "CoreVideo",
  "CoreHaptics",
  "MyFramework"
]
```

Note: `/System/Library/Frameworks` and `/Library/Frameworks` are implicitly added.

**Comment by dimitre** [comment]
I've built this get-chalet github actions, so hopefully I can use to build libraries and pack them in a zip file, multiplatform
https://github.com/dimitre/get-chalet/
it seems to have complete building on macos, some issues on Ubuntu
```
/home/runner/work/get-chalet/get-chalet/chalet/src/Utility/Version.cpp:29:19: error: comparison of integer expressions of different signedness: ‘std::vector<std::__cxx11::basic_string<char> >::size_type’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
[232](https://github.com/dimitre/get-chalet/actions/runs/3222691601/jobs/5272024057#step:2:236)
   29 |  if (split.size() > std::numeric_limits<int>::max())
```

**Reply by rewrking** [reply]
Nice! I'll check this out. Feel free to start a new discussion thread for something like this too. The compiler error you mentioned has been fixed - I'll have a new patch version out shortly (v0.5.6).

**Reply by dimitre** [reply]
Great! I'll be testing right now as I'm cloning directly from git. Yes I'm still unsure about where to write here, discussions vs issues :)

**Reply by rewrking** [reply]
Well, that compile bug would be an issue, while sharing your github action could be a new discussion thread.

**Reply by dimitre** [reply]
Compiled error fixed, building great on Ubuntu now.
